pkg/scheduler/allocator.go

```diff
@@ -72,7 +72,10 @@ func (m *Scheduler) singleStepSchedule(nAlloc int, preemptionParam *preemptionPa
                 proposal := newSingleAllocationProposal(alloc)
                 err := m.updateSchedulingRequestPendingAskByDelta(proposal.AllocationProposals[0], -1)
                 if err == nil {
-                    m.eventHandlers.CacheEventHandler.HandleEvent(newSingleAllocationProposal(alloc))
+                    // update unconfirmed resource of related node for new allocation proposal
+                    alloc.Node.NodeInfo.AddUnconfirmedResource(alloc.SchedulingAsk.AllocatedResource)
```

Comment:
I think we should update _pendingAllocateResource_ of related node immediately after generating an allocation proposal, in that way we can guarantee an exact fairness of nodes, plan B is not to affect cache layer but to keep _pendingAllocateResources_ (structure like map[NodeId]*resources.Resource) inside scheduler layer such as allocator.go, we can join _pendingAllocateResources_ when creating new scheduling nodes for every scheduling process. A better way is to correlate scheduling node with node info, so that we can not create scheduling nodes for every scheduling process and just directly add _pendingAllocateResource_ into scheduling node. Thoughts?

Reply:
The reason to have a cache layer is to support multi-routine scheduling in the future, let's stay with that.  For the issue you mentioned, it probably won't be too bad. Have you tried this way? We can run the same test again to see if it is OK.

Reply:
Currently available resource of node is already deducted in ClusterInfo#processAllocationProposalEvent (further calling stack is PartitionInfo#addNewAllocation --> PartitionInfo#addNewAllocationInternal --> NodeInfo#addAllocation), so I think the first test results can prove this issue, am I right?

Reply:
Hi @TaoYang526 
Got what you are saying, I just want to make sure we are using the best solution to fix this. This is pretty critical.
So can we keep this in `SchedulerNode`, I noticed we already keep tracking of `AllocatingResource` in scheduler node, can we sort nodes with consideration of this part? We can change how to calculate the `CachedAvailableResource` in scheduler node, currently, it is 

```
info.GetAvailableResource()
```

it fetches the latest available resource from the cache layer. Instead, we can initiate an `AvailableResource` in the scheduler node with the initial value `info.GetAvailableResource()`, and then when fetching the latest value during sorting nodes, we do

```
AvailableResource = AvailableResource - AllocatingResource
```

would this work?
This way, we are not holding read lock from node_info (the cache layer), and in each scheduling interval, we maintain and update the nodes info in scheduler nodes.

Please let me know your thought.
Thanks

Reply:
Hi, @yangwwei 
Currently _AllocatingResource_  is only valuable for batch allocations which haven't been supported in once scheduling process, if we can correlate scheduling node with node info, I think it's better enough to take _AllocatingResource_ as _ pendingAllocateResource_, moreover we don't have to create new scheduling nodes for every scheduling process. Thoughts?